### DISCUSSION TITLE
Generate different types of cards using the same ruby file

### DISCUSSION THREAD

**Discussion by LouisDeconinck** [andymeneely/squib#357] [Q&A] [answered]
In my card game, I have multiple types of cards with a totally different layout. I could generate these using new Ruby files, but then these cards would not be generated on the same PDF.

Is there a way to generate different types of cards from the same file? If all cards come from the same .csv file, then defining the number of cards that need to be generated with Squib::Deck.new cards: data['ID'].size becomes hard. I could also use different .csv files for each new card type, but no idea how to use multiple data files in one Ruby file.

Any possibility of generating different types of cards from the same file?

**Comment by andymeneely** [comment, marked as answer]
Is stitching them all into one PDF the only reason to do them in one deck? If so, you can do something like this:

* Say you have Character Attack cards and Character Defend cards. Totally different layouts, but you want them all in the same PDF. 
* Make an `attack.rb` and `defend.rb` separately, they output to things like `_output/char_attack_00.png` and `_output/char_defend_00.png`
* Make a new meta-deck called, say, `pnp.rb`
* Use `Dir['_output/char_*.png']` to get all of those files at once.
* Then make a new deck with all of those cards in it, where the deck is essentially just a single `png` command. 

So essentially the output of one Squib run becomes the input to another. I did something like this here: https://github.com/andymeneely/project-bolt-rats/blob/master/lib/pnp.rb

My rule of thumb is one deck per "sheet" in Excel. If I have everything in one sheet with most of the columns filled out, that's one deck. If I have two sheets with different columns, that's two decks. 

If you _really_ want everything in a single deck, then your best bet is to use `range` on most of your commands to limit which cards they write to. Works great, but gets unwieldy to maintain.

**Reply by LouisDeconinck** [reply]
I have a folder 'individual-cards', this folder contains all of the individual cards I want to merge and only those, in .png format. The cards are generated there from two separate ruby files that output the cards as 'virus_cards-*.png' and 'infection_cards-*.png'. How to merge all of these cards onto one file.

What would be the best way of merging all of these cards onto one pdf file to print?

I have the following code:
```
require 'squib'

Squib::Deck.new(width: 750 * 3, height: 1050 * 3, cards: 2) do
  background color: :white
  Dir['individual-cards/virus_cards-*.png'].to_a.sort.each.with_index do |f,i|
   png file: "../#{f}",
    x: 750 * (i % 3),
    y: 1050 * (i / 3),
    crop_x: 37.5, crop_y: 37.5, crop_width: '2.5in', crop_height: '3.5in',
    range: 0
  end

  Dir['individual-cards/infection_cards-*.png'].to_a.sort.each.with_index do |f,i|
   png file: "../#{f}",
    x: 750 * (i % 3),
    y: 1050 * (i / 3),
    crop_x: 37.5, crop_y: 37.5, crop_width: '2.5in', crop_height: '3.5in',
    range: 1
  end

  save_pdf file: 'pnp_a4.pdf', width: '210mm', height: '297mm'
end
```

This returns the following error:
```
C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/graphics/save_pdf.rb:65:in `initialize': error while writing to output stream (Cairo::WriteError)
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/graphics/save_pdf.rb:65:in `new'   
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/graphics/save_pdf.rb:65:in `init_cc'
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/graphics/save_pdf.rb:14:in `render_pdf'
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/dsl/save_pdf.rb:41:in `run'        
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/dsl/save_pdf.rb:9:in `save_pdf'    
        from merge.rb:21:in `block in <main>'
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/deck.rb:75:in `instance_eval'      
        from C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/squib-0.18.0/lib/squib/deck.rb:75:in `initialize'
        from merge.rb:3:in `new'
        from merge.rb:3:in `<main>'
```

**Reply by LouisDeconinck** [reply]
The error I got was one that happened whenever I tried to generate a new PDF file while the same PDF file was already open. Be sure to close your PDF's as I was not able to alter an opened PDF file.

I also found a cleaner solution for merging all of the cards. The issue with the previous code is that you're limited to only two sheets and each type of card starts on a new sheet. The following code automatically detects the number of pages needed and puts all cards next to each other instead of starting a new page with each new type of card. It does this by creating sub-arrays of 9 cards for each new page.

```
require 'squib'

Squib::Deck.new(width: 750 * 3, height: 1050 * 3 + 3, cards: (Dir['individual-cards/**/*'].length.to_f / 9).ceil ) do
  background color: :white
  allcards = Dir['individual-cards/virus_card-*.png'].to_a.sort + Dir['individual-cards/infection_card-*.png'].to_a.sort
  allcards.each_slice(9).with_index do |ninecards, page|
    ninecards.each.with_index do |f,i|
    png file: "#{f}",
      x: 750 * (i % 3),
      y: 1050 * (i / 3),
      range: page
    end
  end

  save_pdf file: 'pnp_a4.pdf', width: '210mm', height: '297mm'
end
```
